src/main/scala/ohnosequences/loquat/dataProcessing.scala

```diff
@@ -33,90 +36,71 @@ trait AnyProcessingContext {
-
-  /* This is a cover-method, which will be used in the worker run-loop */
-  final def processFiles(
-    dataMappingId: String,
-    inputFilesMap: Map[String, File],
-    workingDir: File
-  ): Result[Map[String, File]] = {
-
-    /* This method serialises OutputFiles data mapping to a normal Map */
-    def filesMap(filesSet: OutputFiles): Map[String, File] =
-      (input.keys.types.asList.map{ _.label }) zip
-      (filesSet.asList.map { _.value.location }) toMap
-
-    parseInputFiles(inputFilesMap mapValues { f => FileDataLocation(f) }) match {
+  val parseInputFiles: AnyApp1At[
+    FileDataLocation ParseDenotations Input#Keys,
+    Map[String,FileDataLocation]
+  ] { type Y = Either[ParseDenotationsError, FileLocationsFor[Input]] }
+
+  type OutputContext = (Output, FileLocationsFor[Output])
+
+  // this is where you define what to do
+  def process(context: ProcessingContext[Input]): Instructions[OutputContext]
+
+  final def runProcess(workingDir: File, inputFiles: Map[String,File]): Result[Map[String, File]] = {
+
+    // TODO move to utils
+    def outputAsMap(outCtx: OutputContext): Map[String, File] =
+      (outCtx._1.keys.types.asList.map{ _.label }) zip
+      (outCtx._2.asList.map { _.value.location }) toMap
+
+    parseInputFiles(inputFiles mapValues { f => FileDataLocation(f) }) match {
       case Left(err) => Failure(err.toString)
       case Right(inputFiles) => {
-        processData(
-          dataMappingId,
-          ProcessingContext[Input, InputFiles](input, inputFiles, workingDir)
+        process(
+          ProcessingContext[Input](inputFiles, workingDir)
         ).run(workingDir.toJava) match {
           case Failure(tr) => Failure(tr)
-          case Success(tr, of) => Success(tr, filesMap(of))
+          case Success(tr, of) => Success(tr, outputAsMap(of))
         }
       }
     }
   }
-}
 
-// case object AnyDataProcessingBundle {
-//
-//   implicit def parseFromFileDataLocation[D <: AnyData](implicit data: D): DenotationParser[D, FileDataLocation, FileDataLocation] =
-//     new DenotationParser(data, data.label)({ f => Some(f) })
-// }
+}
 
 abstract class DataProcessingBundle[
   I <: AnyDataSet,
   O <: AnyDataSet
-](deps: AnyBundle*)(
-  val input: I,
-  val output: O
-)(implicit
-  val parseInputFiles: AnyApp1At[FileDataLocation ParseDenotations I#Keys, Map[String,FileDataLocation]] { type Y = Either[ParseDenotationsError, I#Raw] }
+](deps: AnyBundle*)(implicit
+  val parseInputFiles: AnyApp1At[
+    FileDataLocation ParseDenotations I#Keys,
+    Map[String, FileDataLocation]
+  ] { type Y = Either[ParseDenotationsError, FileLocationsFor[I]] }
 ) extends Bundle(deps: _*) with AnyDataProcessingBundle {
 
   type Input = I
-  // type InputFiles = I#Raw // with AnyKList.withBound[AnyDenotation { type Value = FileDataLocation }]
   type Output = O
-  // type OutputFiles =  O#Raw with AnyKList.withBound[AnyDenotation { type Value = FileDataLocation }]
+
+  // what's the point of instructions for this?
```

Comment:
Fine. But I think it more reasonable to separate that and write a bundle doing it